File: Smallest distinct window - GFG/smallest-distinct-window.java
// { Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;

class GFG {
	public static void main(String[] args) throws IOException
	{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim());
        while(t-->0)
        {
            String str = br.readLine();
            
            Solution obj = new Solution();
            System.out.println(obj.findSubString(str));
            
        }
	}
}

// } Driver Code Ends


//User function Template for Java

class Solution {
    public int findSubString( String str) {
        HashMap<Character, Integer> hm = new HashMap<>();
        for (int i = 0; i < str.length(); i++)
            hm.put(str.charAt(i), 1);
        int i = 0, j = 0, res = Integer.MAX_VALUE, count = hm.size();
        while (j < str.length()) {
            char ch = str.charAt(j);
            hm.put(ch, hm.get(ch) - 1);
            if (hm.get(ch) == 0)
                count--;
            if (count == 0) {
                while (i < str.length() && count == 0) {
                    res = Math.min(res, j - i + 1);
                    hm.put(str.charAt(i), hm.get(str.charAt(i)) + 1);
                    if (hm.get(str.charAt(i)) > 0)
                        count++;
                    i++;
                }
            }
            j++;
        }
        return res;
    }
}

File: Smallest distinct window - GFG/smallest-distinct-window.cpp
// { Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


 // } Driver Code Ends
class Solution{
    public:
    int findSubString(string str)
    {
        // Your code goes here   
        unordered_map<char,int> m;
        
        int dist_count = 0;
        
        for(auto a:str){
            
            if(m.find(a)==m.end()){
                dist_count++;
            }
            m[a]=1;
        }
        
        int i=0;
        int ans=INT_MAX;
        int j=0;
        
        while(i<str.size()){
            
            m[str[i]]--;
            
            if(m[str[i]]==0){
                dist_count--;
            }
            
            if(dist_count==0){
                while(i<str.size() && dist_count==0){
                    
                    ans=min(ans,i-j+1);
                    m[str[j]]++;
                    
                    if(m[str[j]]>0){
                        dist_count++;
                    }
                    
                    j++;
                    
                    
                }
            }
            
            
            i++;
        }
        
        return ans;
        
        
    }
};

// { Driver Code Starts.
// Driver code
int main() {
    int t;
    cin >> t;
    while (t--) {

        string str;
        cin >> str;
        Solution ob;
        cout << ob.findSubString(str) << endl;
    }
    return 0;
}  // } Driver Code Ends